### DISCUSSION TITLE
LinuxBrew:  issue with installing dependencies, glibc@2.13 and gcc@5

### DISCUSSION THREAD

**Discussion by akhst7** [Homebrew/discussions#3399] [Linux]
I've been trying to install several packages that require glibc@2.13 and gcc@5  as dependencies in RHEL7 which is already loaded backward compatible glibc 2.17, and run into issue compiling  glibc@2.13 as follows;

```
==> Upgrading libxml2
  2.9.13 -> 2.9.14_2 

==> Installing dependencies for libxml2: glibc@2.13, gcc@5 and icu4c
==> Installing libxml2 dependency: glibc@2.13
==> ../configure --disable-debug --disable-dependency-tracking --disable-silent-rules --prefix=/geode2/home/u010/aho
Last 15 lines from /N/u/me/Carbonate/.cache/Homebrew/Logs/glibc@2.13/01.configure:
../configure
--disable-debug
--disable-dependency-tracking
--disable-silent-rules
--prefix=/geode2/home/u010/me/Carbonate/.linuxbrew/Cellar/glibc@2.13/2.13
--enable-obsolete-rpc
--without-selinux
--with-headers=/geode2/home/u010/me/Carbonate/.linuxbrew/opt/linux-headers@4.4/include

checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking for suffix of object files... configure: error: in `/tmp/glibcA2.13-20220629-8492-va29e8/glibc-2.13/build':
configure: error: cannot compute suffix of object files: cannot compile

```
Any ways to tell brew to use glibc 2.17 instead of compiling glibc@2.13 ? 

Thanks.

**Comment by zc62** [comment]
I'm having a similar issue. I'm on CentOS 7. When upgrading, brew suddenly asks for glibc2.13 (brew has gcc-5 and glibc2.23), and the compilation failed at
```
{standard input}: Assembler messages:
{standard input}: Error: `_obstack@GLIBC_2.2.5' can't be versioned to common symbol '_obstack_compat'
```

In addition, brew deleted the links to older versions of the programs, whose new versions need glibc2.13, thus breaking a lot of things. (Pretty confused about why glibc2.13 is needed)

EDIT: packages in trouble were libxml2, curl, git and neovim. Running `brew link` restored those packages to older versions.

**Comment by carlocab** [comment]
@Homebrew/linux

**Comment by danielnachun** [comment]
Only two packages have `glibc@2.13` as a dependency - `binutils` and `zlib`.  And both are relocatable and shouldn't need to be built from source by any user.  I've just upgraded my own CentOS 7 install and this did not happen so I'm unsure how this would happen.  Furthermore `zlib` didn't even get a revision bump so it shouldn't even be changed.  `binutils` was revision bumped but it is should need `glibc@2.13` just to install.

**Comment by danielnachun** [comment]
Are these fresh installs or upgrades to existing ones?  That will help us narrow down what's going wrong.

**Reply by zc62** [reply]
In my case these are upgrades.
```
==> Upgrading 4 outdated packages:
libxml2 2.9.14 -> 2.9.14_2
curl 7.83.1 -> 7.84.0
git 2.36.1 -> 2.37.0
neovim 0.7.0 -> 0.7.2
Warning: Building glibc@2.13 from source as the bottle needs:
```

**Reply by akhst7** [reply]
Some of mine are fresh install.

**Reply by danielnachun** [reply]
Okay, for the ones which are **not** fresh installs, can you try to do:
```
brew uninstall --ignore-dependencies zlib; brew install -v zlib
brew uninstall --ignore-dependencies binutils; brew install -v binutils
```
Neither of those should try to compile from source.  After that is done, try to run your upgrades again.

**Reply by zc62** [reply]
It's complicated. My brew relies on brewed git and curl, uninstalling zlib breaks them, therefore I need to go back to compile git by myself. Also if I ignore those errors, I do still seem to see glibc2.13 is being compiled. Will update more after I compile my own git.

**Reply by danielnachun** [reply]
It might be easier to use `miniconda` to get your own copy of `git` and `curl`, but that's up to you.  The most important thing to know here is if `brew install zlib` or `brew install binutils` are falling back to building from source.  `glibc@2.13` is a build-only dependency that should be ignored when installing from a bottle, so either `zlib` or `binutils` are building from source even though they shouldn't, or `brew` is mistakenly trying to install a build-only dependency while installing from a bottle.

**Reply by zc62** [reply]
Oh, I was stupid. I have anaconda3 installed but never realized it ships git and curl. Never mind, I got my git compiled and ran those commands. zlib and binutils install fine without asking for glibc2.13, but when I upgrade again, the same thing happens.

**Comment by Bo98** [comment]
I've found the cause and am able to reproduce the issue. I'm working on a fix.

**Reply by Bo98** [reply]
https://github.com/Homebrew/brew/pull/13494

**Comment by iMichka** [comment]
https://github.com/Homebrew/brew/pull/13494 has been merged; you can now `brew update` and try again.

**Reply by zc62** [reply]
I can confirm it works now.

**Comment by akhst7** [comment]
It works now.  Thanks.

**Reply by wscott** [reply]
I fixed the '_obstack_compat' failure with the following tweak.

```
diff --git a/Formula/glibc@2.13.rb b/Formula/glibc@2.13.rb
index 2aad2f76c19..a9dec355e7c 100644
--- a/Formula/glibc@2.13.rb
+++ b/Formula/glibc@2.13.rb
@@ -114,6 +114,8 @@ class GlibcAT213 < Formula
     # Use brewed ld.so.preload rather than the host's /etc/ld.so.preload
     inreplace "elf/rtld.c", '= "/etc/ld.so.preload";', "= \"#{prefix}/etc/ld.so.preload\";"
 
+    system "perl", "-i", "-lpe", "s/obstack_compat;/obstack_compat = 0;/", "malloc/obstack.c"
+
     mkdir "build" do
       args = [
         "--disable-debug",
```

But the after the recent 2.13_1 change my compiles are again failing:
````
Makefile:167: recipe for target '/tmp/glibcA2.13-20220731-13429-c4smf4/glibc-2.13/build/sunrpc/xbootparam_prot.stmp' failed
Inconsistency detected by ld.so: dynamic-link.h: 209: elf_get_dynamic_info: Assertion `info[15] == ((void *)0)' failed!
make[2]: *** [/tmp/glibcA2.13-20220731-13429-c4smf4/glibc-2.13/build/sunrpc/xbootparam_prot.stmp] Error 127
make[2]: *** Waiting for unfinished jobs....
````

Not sure how to fix that error.  This is on SLES12

**Reply by Bo98** [reply]
You should never need to install glibc@2.13. Please provide more information, like `brew config` and the full output of the command you are running.

**Comment by wscott** [comment]
Huh.  That's odd. A 'brew remove glibc@2.13' worked just fine without complaining about dependencies. Did that change recently? I am sure I had programs that required that package within the last month.

Thanks.

**Reply by Bo98** [reply]
Yes, this was a bug which was fixed in Homebrew 3.5.4.
